### PR TITLE
Reduce duplicate code in Image open

### PR DESCRIPTION
**Pull request by radarhere**
Minor cleanup. If anyone feels if this is more confusing than helpful, feel free to close.

### CODE REVIEW

**Comment by hugovk** [comment]
Note: GitHub is under a DDoS attack:

https://status.github.com/messages

Hence the AppVeyor failures:

```
fatal: unable to access 'https://github.com/python-pillow/Pillow.git/': Failed connect to github.com:443; No error
Command exited with code 128
```

We'll have to restart those later.

**Comment by radarhere** [comment]
The tests have completed now. Coveralls has equated the fewer lines of code with fewer lines of covered code, and so reported a drop in coverage.

File: PIL/Image.py [hunk 1]
```diff
@@ -2299,29 +2299,18 @@ def open(fp, mode="r"):
...
-                factory, accept = OPEN[i]
-                if not accept or accept(prefix):
-                    fp.seek(0)
-                    im = factory(fp, filename)
-                    _decompression_bomb_check(im.size)
-                    return im
-            except (SyntaxError, IndexError, TypeError, struct.error):
-                logger.debug("", exc_info=True)
+    for i in range(0, 2):
+        if i == 0 or init():
```
Comment (wiredfool) [inline PIL/Image.py]: Decent idea to dedup code, but this looks legally confusing.

Comment (radarhere) [inline PIL/Image.py]: I know. I can't seem to think of a neat solution. Again, feel free to close if this just doesn't work.

Comment (hugovk) [inline PIL/Image.py]: How about something like:
```python
def _something(prefix, fp, filename):
    for i in ID:
        try:
            factory, accept = OPEN[i]
            if not accept or accept(prefix):
                fp.seek(0)
                im = factory(fp, filename)
                _decompression_bomb_check(im.size)
                return im
        except (SyntaxError, IndexError, TypeError, struct.error):
            logger.debug("", exc_info=True)


def open(fp, mode="r"):
    ...
    ...
    ...
    _something(prefix, fp, filename)

    if init():

        _something(prefix, fp, filename)

    raise IOError("cannot identify image file %r"
                  % (filename if filename else fp))
```

Comment (wiredfool) [inline PIL/Image.py]: I've got this passing the tests:

```
def open(fp, mode="r"):
    """
    Opens and identifies the given image file.

    This is a lazy operation; this function identifies the file, but
    the file remains open and the actual image data is not read from
    the file until you try to process the data (or call the
    :py:meth:`~PIL.Image.Image.load` method).  See
    :py:func:`~PIL.Image.new`.

    :param fp: A filename (string), pathlib.Path object or a file object.
       The file object must implement :py:meth:`~file.read`,
       :py:meth:`~file.seek`, and :py:meth:`~file.tell` methods,
       and be opened in binary mode.
    :param mode: The mode.  If given, this argument must be "r".
    :returns: An :py:class:`~PIL.Image.Image` object.
    :exception IOError: If the file cannot be found, or the image cannot be
       opened and identified.
    """

    if mode != "r":
        raise ValueError("bad mode %r" % mode)

    filename = ""
    if isPath(fp):
        filename = fp
    elif sys.version_info >= (3, 4):
        from pathlib import Path
        if isinstance(fp, Path):
            filename = str(fp.resolve())
    if filename:
        fp = builtins.open(filename, "rb")

    try:
        fp.seek(0)
    except (AttributeError, io.UnsupportedOperation):
        fp = io.BytesIO(fp.read())

    prefix = fp.read(16)

    preinit()

    def _open_core(fp, filename, prefix):
        for i in ID:
            try:
                factory, accept = OPEN[i]
                if not accept or accept(prefix):
                    fp.seek(0)
                    im = factory(fp, filename)
                    _decompression_bomb_check(im.size)
                    return im
            except (SyntaxError, IndexError, TypeError, struct.error):
                logger.debug("", exc_info=True)
        return None

    im = _open_core(fp, filename, prefix)
    
    if im is None:
        if init():
            im = _open_core(fp, filename, prefix)

    if im:
        return im

    raise IOError("cannot identify image file %r"
                  % (filename if filename else fp))
```

Comment (radarhere) [inline PIL/Image.py]: Thanks! For some reason, I thought that the option of separating it out into a function would look more garbled than that. Nice work.

**Comment by radarhere** [comment]
Closed in favour of #1415

**Pull request closed without merging**